include/circt/Dialect/FIRRTL/FIRRTLAttributes.h

```diff
@@ -14,6 +14,110 @@
 #define CIRCT_DIALECT_FIRRTL_FIRRTLATTRIBUTES_H
 
 #include "circt/Dialect/FIRRTL/FIRRTLDialect.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/ADT/iterator.h"
+
+namespace circt {
+namespace firrtl {
+
+//===----------------------------------------------------------------------===//
+// PortDirections
+//===----------------------------------------------------------------------===//
+
+/// This represents the direction of a single port.
+enum class Direction { Input = 0, Output };
+
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+                                     const Direction &dir) {
+  return os << (dir == Direction::Input ? "in" : "out");
+}
+
+/// This represents the directions of ports in a module.  This is a simple type
+/// safe wrapper around a BitVector.
+class PortDirections {
+public:
+  /// Iterator for PortDirections returning Directions.
+  struct Iterator
+      : public llvm::iterator_facade_base<Iterator, std::forward_iterator_tag,
+                                          Direction> {
+    explicit Iterator(const PortDirections &directions, size_t index = 0)
+        : directions(&directions), index(index) {}
+    Iterator &operator=(const Iterator &) = default;
+    using llvm::iterator_facade_base<Iterator, std::forward_iterator_tag,
+                                     Direction>::operator++;
+    Iterator &operator++() { return ++index, *this; }
+    Direction operator*() const { return directions->at(index); }
+    bool operator==(const Iterator &rhs) const { return index == rhs.index; }
+    bool operator!=(const Iterator &rhs) const { return index != rhs.index; }
+
+  private:
+    const PortDirections *directions;
+    size_t index;
+  };
+
+  using value_type = Direction;
+  using iterator = Iterator;
+
+  /// Construct an empty PortDirections.
+  PortDirections() = default;
+
+  /// Construct a PortDirections from a array of Directions.
+  PortDirections(ArrayRef<Direction> directions) {
+    reserve(directions.size());
+    llvm::copy(directions, std::back_inserter(*this));
+  }
+
+  bool operator==(const PortDirections &rhs) const {
+    return directions == rhs.directions;
+  }
+
+  /// Get the direction of a specifc port.
+  Direction at(unsigned index) const {
+    unsigned val = directions[index];
+    return static_cast<Direction>(val);
```

Comment:
Should the `val` just be inlined into the `static_cast`?

Reply:
Woops I means to clean this up.  You might find this bit of c++ interesting: BitVector has [two implementations](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/BitVector.h#L436) of `operator[]`, one of which returns a mutable reference to a specific bit.  The fact that there were two implementations made my code here a bit weird and I forgot to fix it.